### PR TITLE
tests(utils): move tests to TS

### PR DESCRIPTION
**Pull request by hemal7735**
**What kind of change does this PR introduce?**

We are moving `utils` package test-cases to TS.

**Did you add tests for your changes?** NA

**If relevant, did you update the documentation?** NA

**Summary**

This PR address the #613 - tests(migration): typescript

**Does this PR introduce a breaking change?** NA

# Other information
## There are some tests files for which help is needed, I have changed them to `.tx` to allow me to push to GitHub.

**Pending Tasks:**
- [x]  npm-packages-exists.test.tx - TS mock issue - `npm-exists` module is not getting mocked, as actual call is going 
- [x]  recursive-parser.test.tx -- defineTest parameter issue , already mentioned in #613 
- [x] cleanup equivalent `.js` , `__fixtures__` and `__snapshots__` files. (these are not removed for reference)

### CODE REVIEW

**Comment by hemal7735** [comment]
@ematipico can you help moving things forwards?

**Comment by hemal7735** [comment]
@dhruvdutt can you help me with the blockers?

**Comment by ematipico** [comment]
Could please some comment on your code describing the problem in a detailed way?

File: packages/utils/__tests__/recursive-parser.test.tx [hunk 1]
```diff
@@ -0,0 +1,64 @@
+"use strict";
+
+import defineTest from "../defineTest";
+
+defineTest(__dirname, "init", "fixture-1", "entry", {
```
Comment (ematipico) [inline packages/utils/__tests__/recursive-parser.test.tx]: Have you tried to extend the types of defineTest ?

Comment (ematipico) [reply packages/utils/__tests__/recursive-parser.test.tx]: Trying to expand the current types so the function is able to accept also other primitives.

File: packages/utils/__tests__/npm-packages-exists.test.tx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import npmPackagesExists from "../npm-packages-exists";
+// console.log('npmPackagesExists: ', npmPackagesExists);
+import {resolvePackages} from "../resolve-packages";
+
+jest.mock("../npm-exists");
```
[the remark this answers is not in the document]
Comment (ematipico) [reply packages/utils/__tests__/npm-packages-exists.test.tx]: 1. Or better solution, doing
```js
(Mocked as any). whatever
```
We don't really care about the types. But surely type casting is a better solution. If it work, please use it

2. I really don't know why :(

File: packages/utils/__tests__/recursive-parser.test.tx (hunk 1 above)
Comment (hemal7735) [reply packages/utils/__tests__/recursive-parser.test.tx]: that helped passing test-cases, but don't we need to chagne the description as well. Param says `action`, and we are passing an `object`.
`action init, update or remove, decides how to format the AST` this is what it says.

Comment (evenstensberg) [reply packages/utils/__tests__/recursive-parser.test.tx]: This func takes a name and extracts a testcase out of the params it get's. Code should be readable, and that is correct @hemal7735

Comment (hemal7735) [reply packages/utils/__tests__/recursive-parser.test.tx]: Okay, so I will go ahead and add more types to pass the test cases.
But we will need an issue to track if we really want to refactor this function?

File: packages/utils/__tests__/npm-packages-exists.test.tx (hunk 2 above)
[the remark this answers is not in the document]
Comment (hemal7735) [reply packages/utils/__tests__/npm-packages-exists.test.tx]: @ematipico I did some debugging to trace the path of the code execution. The error was thrown during compile/build time from jest. Whatever we trying to mock is being mocked. 😃

File: packages/utils/__tests__/ast-utils.test.ts [hunk 3]
```diff
@@ -105,41 +105,41 @@ const a = { plugs: [] }
 	describe("createOrUpdatePluginByName", () => {
 		it("should create a new plugin without arguments", () => {
 			const ast = j("{ plugins: [] }");
-			ast.find(j.ArrayExpression).forEach(node => {
-				utils.createOrUpdatePluginByName(j, node, "Plugin");
+			ast.find(j.ArrayExpression).forEach((node) => {
```
Comment (evenstensberg) [inline packages/utils/__tests__/ast-utils.test.ts]: @ematipico I think we were agreeing on non-enclosed paran when single arg?

Comment (hemal7735) [reply packages/utils/__tests__/ast-utils.test.ts]: Current `tslint` configuration is strict in those areas. `eslint` config is still relaxed.

Comment (evenstensberg) [reply packages/utils/__tests__/ast-utils.test.ts]: Can we make it as we agreed? Think Emanuelle has some comments there

Comment (hemal7735) [reply packages/utils/__tests__/ast-utils.test.ts]: We will need to use "arrow-parens": [true, "ban-single-arg-parens"]" in tslint config 
more info: https://palantir.github.io/tslint/rules/arrow-parens/

File: packages/utils/__tests__/npm-packages-exists.test.ts [hunk 4]
```diff
@@ -1,10 +1,12 @@
-const npmPackagesExists = require("./npm-packages-exists").default;
+import npmPackagesExists from "../npm-packages-exists";
 // console.log('npmPackagesExists: ', npmPackagesExists);
```
Comment (dhruvdutt) [inline packages/utils/__tests__/npm-packages-exists.test.ts]: You can clean this.

File: packages/utils/__tests__/ast-utils.test.ts [hunk 5]
```diff
@@ -105,41 +105,41 @@ const a = { plugs: [] }
...
+				utils.createOrUpdatePluginByName(j, node, "Plugin", {});
```
Comment (dhruvdutt) [inline packages/utils/__tests__/ast-utils.test.ts]: I think you can make `options` as optional param and clean this up.

File: packages/utils/__tests__/__snapshots__/ast-utils.test.ts.snap [hunk 6]
```diff
@@ -59,7 +57,7 @@ exports[`utils createOrUpdatePluginByName should create a new plugin with argume
 })] }"
 `;
 
-exports[`utils createOrUpdatePluginByName should create a new plugin without arguments 1`] = `"{ plugins: [new Plugin()] }"`;
+exports[`utils createOrUpdatePluginByName should create a new plugin without arguments 1`] = `"{ plugins: [new Plugin({})] }"`;
```
Comment (dhruvdutt) [inline packages/utils/__tests__/__snapshots__/ast-utils.test.ts.snap]: This will get cleaned if you make `options` as optional param.

**Comment by hemal7735** [comment]
> Ts thing on param needs to be sorted out

@evenstensberg are you talking about parenthesis from previous conversation?

**Comment by hemal7735** [comment]
>Add INode type for all node paths.

@dhruvdutt can you comment on the file where you think change is required?

File: packages/utils/__tests__/ast-utils.test.ts (hunk 3 above)
Comment (dhruvdutt) [reply packages/utils/__tests__/ast-utils.test.ts]: ~`forEach((node) => {`~ `forEach((node: INode) => {`

@hemal7735 Do this for all occurrences of `node` and `path`. Check `ast-utils.ts` files for reference.

**Comment by ematipico** [comment]
Resolving the TS linter problem will cause a waterfall effect where other files will be affected by the change. We can make a separate PR to fix the linter and we keep focusing on migrating the tests

**Comment by dhruvdutt** [comment]
Agreed. I'm working on removing tslint. We can merge this for now.

**Comment by dhruvdutt** [comment]
Please fix conflicts.

**Comment by hemal7735** [comment]
>Please fix conflicts.

@dhruvdutt I don't see github showing me any conflict.

**Pull request merged**